### ISSUE TITLE
Function node editor not supporting valid switch syntax (multi-case - chained operations)

### ISSUE THREAD

**Issue by TheOriginalMrWolf** [node-red/node-red#1911]
Hi,

I tried writing a function node with valid javascript switch syntax for multi-case chained operations (as per last example at: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch) and the editor shows errors (red x box) with the text: Expected a 'break' statement before 'case'.  Trying to deploy the function node fails with "The workspace contains some nodes that are not properly configured..." ... "Are you sure you want to deploy?"... and, of course, the node doesn't run.

To be clear, what I'm trying to do is the same as the mozilla example:

```
var foo = 1;
var output = 'Output: ';
switch (foo) {
  case 0:
    output += 'So ';
  case 1:
    output += 'What ';
    output += 'Is ';
  case 2:
    output += 'Your ';
  case 3:
    output += 'Name';
  case 4:
    output += '?';
    console.log(output);
    break;
  case 5:
    output += '!';
    console.log(output);
    break;
  default:
    console.log('Please pick a number from 0 to 5!');
}
```
With the expected result:
```
      foo is NaN or not 1, 2, 3, 4, 5 or 0:    Please pick a number from 0 to 5!
                                         0:    Output: So What Is Your Name?
                                         1:    Output: What Is Your Name?
                                         2:    Output: Your Name?
                                         3:    Output: Name?
                                         4:    Output: ?
                                         5:    Output: !
```

I thought this has been supported for years?

**Comment by knolleary** [comment]
Hi, please provide the extra information that was in the issue template you've deleted:

```
- [ ] Node-RED version:
- [ ] node.js version:
- [ ] npm version:
- [ ] Platform/OS:
- [ ] Browser:
```

When I paste your code into a Function node, I get yellow warning triangles (as forgetting the break statement is a common cause of bugs), but they are not marked as errors.

<img width="338" alt="local_node-red" src="https://user-images.githubusercontent.com/51083/46349609-69d13d00-c64a-11e8-9c73-84a6387f1a93.png">

As the code you've shared is the moz example code, could you provide the node-red specific code you're trying to use?

**Comment by TheOriginalMrWolf** [comment]
My apologies!!

- Node-RED version: v0.19.4
- Node.js  version: v8.12.0
- NPM: 6.4.1
- Linux 4.14.70-v7+ arm LE (Dietpi distro on Raspberry Pi)
- Browser: Google Chrome Version 69.0.3497.100 (Official Build) (64-bit) on Windows 10

Soooo... this is interesting.

I had exactly that problem with the Moz code -- and it blew my mind that you copy/pasted same and only got warnings.

Thinking I'm going nuts, I again created a function node and copy / pasted the Moz example in.

YELLOW TRIANGLES this time.  WTF???

...and then I typed in right at the end
```
return msg;
```

BANG.  RED TRIANGLES...

Whaaa???

**Comment by knolleary** [comment]
Huh - I see what you mean.

We embed the [ace editor](https://ace.c9.io) and are completely reliant on the language support they provide. I've seen some other quirks like this around some of the newer es6 syntax - warnings that become errors if you do certain things.

We can try raising this upstream with ace - but I would not expect a quick turnaround.

**Comment by TheOriginalMrWolf** [comment]
That would be awesome -- thank you!!!!

**Comment by knolleary** [comment]
Figured it out - not an upstream issue, but one for us to fix which I have done for the 0.20 release - whenever that arrives. Happily this addresses all the quirks I mentioned before.

In short, the code you type into the Function node gets wrapped in a Function itself before being run. To properly validate the javascript code, it needs to be examined in its wrapped form - not the raw text you type in. To do that, I've created a custom edit mode for the Function node to use that does the necessary work.
